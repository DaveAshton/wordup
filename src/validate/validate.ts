import { CellData, GameData, LetterStatus } from "../model";

const dictionary = require("./dictionary");

const lookup = new Set(dictionary);

export const isValidWord = (word: string): boolean => {
  const isValid = lookup.has(word.toLocaleLowerCase());
  console.log("isvalid", word, isValid);
  return isValid;
};

const getNextCell = ({ focussed, cells }: GameData): CellData => {
  const { rowId, cellId } = focussed;
  // console.log('getNextCell', cellId, rowId);
  if (cellId < cells[rowId].length - 1) {
    return {
      rowId,
      cellId: cellId + 1,
      status: LetterStatus.NotSet,
    };
  }

  return focussed;
};

const getPrevCell = ({ focussed, cells }: GameData): CellData => {
  const { rowId, cellId } = focussed;
  // console.log('getPrevCell', cellId, rowId);
  if (cellId > 0) {
    return {
      rowId,
      cellId: cellId - 1,
      status: LetterStatus.NotSet,
    };
  }
  return focussed;
};

const validateRow = (
  gameData: GameData,
  gameWord: string,
  isValidWord?: boolean
): GameData => {
  const { focussed, cells } = gameData;
  const { rowId } = focussed;
  const upperWord = gameWord.toLocaleUpperCase();

  const wordMap = upperWord.split("").reduce((map, letter) => {
    let count = map.get(letter);
    return map.set(letter, count ? count+1 : 1);
  }, new Map<string, number>());
  // console.log('word map', wordMap);

  const deleteLetter = (letter: string) => {
    let count = wordMap.get(letter);
    if (count) {
      wordMap.set(letter, count - 1);
    }
  };

  const isCorrectCell = (cell: CellData) => {
    const isCorrectLetter = cell.letter === upperWord[cell.cellId];
    if (isCorrectLetter && cell.letter) {
      deleteLetter(cell.letter);
    }
    return {
      ...cell,
      status: isCorrectLetter ? LetterStatus.Correct : LetterStatus.Incorrect,
    };
  };
  let updatedRow = cells[rowId].map(isCorrectCell);
  updatedRow = updatedRow.map((cell) => {

    if (cell.status === LetterStatus.Correct) {
      return cell;
    }
    if (cell.letter && !!wordMap.get(cell.letter)) {
      deleteLetter(cell.letter);
      return {
        ...cell,
        status: LetterStatus.ValidOutOfPosition,
      };
    } else if (cell.letter && !wordMap.get(cell.letter)) {
      return {
        ...cell,
        status: LetterStatus.Incorrect,
      };
    }
    if (cell.letter) {
      deleteLetter(cell.letter);
    }

    return cell;
  });

  const success = updatedRow.every((cell, cellIdx) => cell.letter === upperWord[cellIdx]);
  const isLastRow = (isValidWord && rowId === cells.length - 1);
  const gameComplete = success || isLastRow;
 
  const ret = {
    ...gameData,
    cells: cells.map((row, rowIdx) => {
      if (rowIdx === rowId) {
        return updatedRow;
      }
      return row;
    }),
    gameComplete,
    success
  };
  console.log("validated", ret.usedLetters, ret.focussed);
  return ret;
};
const getNextRow = (gameData: GameData): GameData => {
  const { focussed, cells } = gameData;
  const { rowId } = focussed;
  // if cellid is end of row and there are more rows available
  if (rowId < cells.length - 1) {
    return {
      ...gameData,
      focussed: {
        rowId: rowId + 1,
        cellId: 0,
        status: LetterStatus.NotSet,
      },
    };
  }
  return gameData;
};
const updateCell = (gameData: GameData, value?: string): GameData => {
  if (gameData.gameComplete) {
    return {
      ...gameData,
      focussed: {
        cellId: -1,
        rowId: -1,
        status: LetterStatus.NotSet,
      },
      rowError: undefined,
    };
  }

  const { focussed, cells } = gameData;
  let updatedFocussed = value ? getNextCell(gameData) : getPrevCell(gameData);
  // console.log('foc', updatedFocussed.cellId, updatedFocussed.rowId);
  const updatedCells = cells.map((row, rowIdx) => {
    if (rowIdx === focussed.rowId) {
      return row.map((cell, cellIdx) => {
        if (cellIdx === focussed.cellId) {
          return {
            ...cell,
            letter: value,
          };
        }
        return cell;
      });
    }
    return row;
  });
  return {
    ...gameData, // new
    cells: updatedCells,
    focussed: updatedFocussed,
    rowError: undefined,
  };
};

 const validateGame = (
  gameData: GameData,
  code: number,
  gameWord: string,
  value?: string
): GameData => {
  // console.log(
  //   ">>",
  //   gameData.focussed.cellId === gameData.cells[0].length - 1,
  //   gameData.focussed.cellId,
  //   gameData.cells[0].length - 1,
  //   gameData.cells[gameData.focussed.rowId][gameData.focussed.cellId].letter
  // );
  if (code > 64 && code < 91) {
    return updateCell(gameData, value?.toLocaleUpperCase());
  } else if (value === "Backspace" || value === "{bksp}") {
    return updateCell(gameData, undefined);
  } else if (value === "Enter" || value === "{enter}") {
    if (
      gameData.focussed.cellId === gameData.cells[0].length - 1 &&
      gameData.cells[gameData.focussed.rowId][gameData.focussed.cellId].letter
    ) {
      if (
        isValidWord(
          gameData.cells[gameData.focussed.rowId].map((x) => x.letter).join("")
        )
      ) {
        const upd = validateRow(gameData, gameWord, true);
        if (upd.gameComplete) {
          return updateCell(upd);
        }
        return getNextRow(upd);
      } else {
        return {
          ...gameData,
          rowError: "Invalid word.",
        };
      }
    }
  }

  return gameData;
};

export const updateGame = (
  gameData: GameData,
  code: number,
  gameWord: string,
  value?: string
): GameData => {
  const dt = validateGame(gameData, code, gameWord, value);
  const {cells, focussed, usedLetters} = dt;

  if (focussed.rowId > 0) {
    cells[focussed.rowId-1].forEach((cell, idx) => {
      if (cell.letter && cell.status === LetterStatus.Correct) {
        usedLetters[LetterStatus.Correct].add(cell.letter);
        if (usedLetters[LetterStatus.ValidOutOfPosition].has(cell.letter)) {
          usedLetters[LetterStatus.ValidOutOfPosition].delete(cell.letter);
        }
  
      } else if (cell.letter && cell.status === LetterStatus.ValidOutOfPosition) {
        if (!usedLetters[LetterStatus.Correct].has(cell.letter)) {
            usedLetters[LetterStatus.ValidOutOfPosition].add(cell.letter);
        }
        
      } else if (cell.letter && cell.status === LetterStatus.Incorrect) {
        if (!usedLetters[LetterStatus.Correct].has(cell.letter) && !usedLetters[LetterStatus.ValidOutOfPosition].has(cell.letter)) {
            usedLetters[LetterStatus.Incorrect].add(cell.letter);
        }
        
      }
    });
  }

  return {
    ...dt,
    usedLetters: {
      ...usedLetters
    }
  };

}


export const createGame = (wordLength: number): GameData => {
 // console.log("creating game", wordLength, lookup.size);
  const cells = createCells(wordLength);
  return {
    focussed: cells[0][0],
    cells,
    usedLetters: {
      [LetterStatus.Correct]: new Set<string>(),
      [LetterStatus.ValidOutOfPosition]: new Set<string>(),
      [LetterStatus.Incorrect]: new Set<string>()
    }
  };
};
const createCells = (wordLength: number): CellData[][] => {
  const res = [];

  for (let rowId = 0; rowId < wordLength + 1; rowId++) {
    res.push(
      Array(wordLength)
        .fill(undefined)
        .map((cl, cellId) => ({ status: LetterStatus.NotSet, cellId, rowId }))
    );
  }
  return res;
};
