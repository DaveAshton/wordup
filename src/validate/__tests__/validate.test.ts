import { GameData } from "..";
import { updateGame, createGame, LetterStatus, CellData } from "../validate";

jest.mock("../dictionary", () => {
  return [
    "swine",
    "strip",
    "witch",
    "bears",
    "winter",
    "sausage",
    "beast",
    "there",
    "trees",
  ];
});

const update =
  (gameWord: string) => (data: GameData, letter: string, code?: number) => {
    const keyCode = code == undefined ? letter.charCodeAt(0) : code;
    // console.log('keycode', keyCode);
    return updateGame(data, keyCode, gameWord, letter);
  };

const updateMultiple = (
  keyedWord: string,
  gameWord: string,
  data: GameData
): GameData => {
  const func = update(gameWord);
  return keyedWord.split("").reduce((dt, letter) => func(dt, letter), data);
};

const expectCell = (
  data: GameData,
  rowIdx: number,
  cellIdx: number,
  letter: string,
  status: LetterStatus
) => {
  expect(data.cells[rowIdx][cellIdx]).toEqual({
    rowId: rowIdx,
    cellId: cellIdx,
    letter,
    status,
  });
};

type Status =
  | LetterStatus.Correct
  | LetterStatus.Incorrect
  | LetterStatus.ValidOutOfPosition;

const expectUsedLetters = (
  data: GameData,
  status: Status,
  expectation: string[]
) => {
  expect(Array.from(data.usedLetters[status].values())).toEqual(expectation);
};

describe("createGame tests", () => {
  let gameData: GameData;
  const wordLength = 5;
  beforeEach(() => {
    gameData = createGame(wordLength);
  });

  it("should generate cells of expected length", () => {
    // rows equals word length + 1
    expect(gameData.cells.length).toEqual(wordLength + 1);
    expect(gameData.cells[0].length).toEqual(wordLength);
  });

  it("should set focussed cell to 0,0", () => {
    expect(gameData.focussed.cellId).toEqual(0);
    expect(gameData.focussed.rowId).toEqual(0);
  });
});

describe("updateGame tests", () => {
  let gameData: GameData;
  const wordLength = 5;
  const gameWord = "bears";
  const upd = update(gameWord);

  beforeEach(() => {
    gameData = createGame(wordLength);
  });

  describe("given user adds invalid letter", () => {
    const keyedLetter = "T";
    beforeEach(() => {
      gameData = upd(gameData, keyedLetter);
    });

    it("should add letter to cells", () => {
      expect(gameData.cells[0][0]).toEqual({
        rowId: 0,
        cellId: 0,
        letter: "T",
        status: LetterStatus.NotSet,
      });
    });

    it("should move focussed cell to 0,1", () => {
      expect(gameData.focussed).toEqual({
        rowId: 0,
        cellId: 1,
        letter: undefined,
        status: LetterStatus.NotSet,
      });
    });
  });

  describe("given user adds valid word but not game word", () => {
    const keyedWord = "BEAST";
    beforeEach(() => {
      gameData = updateMultiple(keyedWord, gameWord, gameData);
    });

    it("should add letter to cells", () => {
      // console.log('gd', keyedLetter.charCodeAt(0), gameData.cells[0][0]);
      expectCell(gameData, 0, 0, "B", LetterStatus.NotSet);
      expectCell(gameData, 0, 1, "E", LetterStatus.NotSet);
      expectCell(gameData, 0, 2, "A", LetterStatus.NotSet);
      expectCell(gameData, 0, 3, "S", LetterStatus.NotSet);
      expectCell(gameData, 0, 4, "T", LetterStatus.NotSet);
    });

    it("should move focussed cell to 0,4", () => {
      expect(gameData.focussed).toEqual({
        rowId: 0,
        cellId: 4,
        letter: undefined,
        status: LetterStatus.NotSet,
      });
    });

    describe("then hits enter", () => {
      beforeEach(() => {
        gameData = upd(gameData, "Enter", 13);
      });

      it("should validate cells in row", () => {
        expectCell(gameData, 0, 0, "B", LetterStatus.Correct);
        expectCell(gameData, 0, 1, "E", LetterStatus.Correct);
        expectCell(gameData, 0, 2, "A", LetterStatus.Correct);
        expectCell(gameData, 0, 3, "S", LetterStatus.ValidOutOfPosition);
        expectCell(gameData, 0, 4, "T", LetterStatus.Incorrect);
      });

      it("should move focussed cell to next row 1,0", () => {
        expect(gameData.focussed).toEqual({
          rowId: 1,
          cellId: 0,
          letter: undefined,
          status: LetterStatus.NotSet,
        });
      });

      it("should have B E A in Correct usedLetters Set", () => {
        expectUsedLetters(gameData, LetterStatus.Correct, ["B", "E", "A"]);
      });

      it("should have S in ValidOutOfPosition usedLetters Set", () => {
        expectUsedLetters(gameData, LetterStatus.ValidOutOfPosition, ["S"]);
      });

      it("should have T in Incorrect usedLetters Set", () => {
        expectUsedLetters(gameData, LetterStatus.Incorrect, ["T"]);
      });

      it("should not be complete", () => {
        expect(gameData.gameComplete).toEqual(false);
      });

      it("should not have row error", () => {
        expect(gameData.rowError).toBeFalsy();
      });
    });
  });

  describe("given user adds invalid word", () => {
    const keyedWord = "PARPS";
    beforeEach(() => {
      gameData = updateMultiple(keyedWord, gameWord, gameData);
    });

    it("should add letter to cells", () => {
      expectCell(gameData, 0, 0, "P", LetterStatus.NotSet);
      expectCell(gameData, 0, 1, "A", LetterStatus.NotSet);
      expectCell(gameData, 0, 2, "R", LetterStatus.NotSet);
      expectCell(gameData, 0, 3, "P", LetterStatus.NotSet);
      expectCell(gameData, 0, 4, "S", LetterStatus.NotSet);
    });

    it("should move focussed cell to 0,4", () => {
      expect(gameData.focussed).toEqual({
        rowId: 0,
        cellId: 4,
        letter: undefined,
        status: LetterStatus.NotSet,
      });
    });

    describe("then hits enter", () => {
      beforeEach(() => {
        gameData = upd(gameData, "Enter", 13);
      });

      it("should NOT move focussed cell to next row (stays at end of row)", () => {
        expect(gameData.focussed).toEqual({
          rowId: 0,
          cellId: 4,
          letter: undefined,
          status: LetterStatus.NotSet,
        });
      });

      it("should have row error", () => {
        expect(gameData.rowError).toBeTruthy();
      });
    });
  });
});

describe("given game word with double letters", () => {
  let gameData: GameData;
  const wordLength = 5;
  const gameWord = "TREES";
  const upd = update(gameWord);

  beforeEach(() => {
    gameData = createGame(wordLength);
  });

  describe("given user adds valid word but not game word", () => {
    const keyedWord = "THERE";
    beforeEach(() => {
      gameData = updateMultiple(keyedWord, gameWord, gameData);
    });

    it("should add letter to cells", () => {
      expectCell(gameData, 0, 0, "T", LetterStatus.NotSet);
      expectCell(gameData, 0, 1, "H", LetterStatus.NotSet);
      expectCell(gameData, 0, 2, "E", LetterStatus.NotSet);
      expectCell(gameData, 0, 3, "R", LetterStatus.NotSet);
      expectCell(gameData, 0, 4, "E", LetterStatus.NotSet);
    });

    it("should move focussed cell to 0,4", () => {
      expect(gameData.focussed).toEqual({
        rowId: 0,
        cellId: 4,
        letter: undefined,
        status: LetterStatus.NotSet,
      });
    });

    describe("then hits enter", () => {
      beforeEach(() => {
        gameData = upd(gameData, "Enter", 13);
      });

      it("should validate cells in row", () => {
        expectCell(gameData, 0, 0, "T", LetterStatus.Correct);
        expectCell(gameData, 0, 1, "H", LetterStatus.Incorrect);
        expectCell(gameData, 0, 2, "E", LetterStatus.Correct);
        expectCell(gameData, 0, 3, "R", LetterStatus.ValidOutOfPosition);
        expectCell(gameData, 0, 4, "E", LetterStatus.ValidOutOfPosition);
      });

      it("should move focussed cell to next row 1,0", () => {
        expect(gameData.focussed).toEqual({
          rowId: 1,
          cellId: 0,
          letter: undefined,
          status: LetterStatus.NotSet,
        });
      });

      it("should have T E in Correct usedLetters Set", () => {
        expectUsedLetters(gameData, LetterStatus.Correct, ["T", "E"] );
      });

      it("should have R E in ValidOutOfPosition usedLetters Set", () => {
        expectUsedLetters(gameData, LetterStatus.ValidOutOfPosition, [ "R"] );
      });

      it("should have H in Incorrect usedLetters Set", () => {
        expectUsedLetters(gameData, LetterStatus.Incorrect, [ "H"] );
      });

      it("should not be complete", () => {
        expect(gameData.gameComplete).toEqual(false);
      });

      it("should not have row error", () => {
        expect(gameData.rowError).toBeFalsy();
      });
    });
  });
});

describe("given game played to completion", () => {
  let gameData: GameData;
  const wordLength = 5;
  const gameWord = "TREES";
  const upd = update(gameWord);

  beforeEach(() => {
    gameData = createGame(wordLength);
  });

  describe("given user adds 2 valid words, then invalid word", () => {
    beforeEach(() => {
      gameData = updateMultiple("SWINE", gameWord, gameData);
      gameData = upd(gameData, "Enter", 13);
      gameData = updateMultiple("STRIP", gameWord, gameData);
      gameData = upd(gameData, "Enter", 13);
      gameData = updateMultiple("PARPS", gameWord, gameData); //INVALID
      gameData = upd(gameData, "Enter", 13);
 
    });

    it("should add letter to cells", () => {
      expectCell(gameData, 2, 0, "P", LetterStatus.NotSet);
      expectCell(gameData, 2, 1, "A", LetterStatus.NotSet);
      expectCell(gameData, 2, 2, "R", LetterStatus.NotSet);
      expectCell(gameData, 2, 3, "P", LetterStatus.NotSet);
      expectCell(gameData, 2, 4, "S", LetterStatus.NotSet);
    });

    it("should have row error", () => {
      expect(gameData.rowError).toBeTruthy();
    });

    describe("then backspaces and enters correct word", () => {
      beforeEach(() => {
        gameData = upd(gameData, "Backspace", 13);
        gameData = upd(gameData, "Backspace", 13);
        gameData = upd(gameData, "Backspace", 13);
        gameData = upd(gameData, "Backspace", 13);
        gameData = upd(gameData, "Backspace", 13);
        gameData = updateMultiple(gameWord, gameWord, gameData);
        gameData = upd(gameData, "Enter", 13);
      });

      it("should have validate row", () => {
        expectCell(gameData, 2, 0, "T", LetterStatus.Correct);
        expectCell(gameData, 2, 1, "R", LetterStatus.Correct);
        expectCell(gameData, 2, 2, "E", LetterStatus.Correct);
        expectCell(gameData, 2, 3, "E", LetterStatus.Correct);
        expectCell(gameData, 2, 4, "S", LetterStatus.Correct);
      });

      it("should be complete", () => {
        expect(gameData.gameComplete).toEqual(true);
      });

      it("should not have row error", () => {
        expect(gameData.rowError).toBeFalsy();
      });
    });
  });
});