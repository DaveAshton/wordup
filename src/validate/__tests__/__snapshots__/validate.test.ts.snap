// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given game played to completion given user adds 2 valid words, then invalid word then backspaces and enters correct word should match snapshot 1`] = `
Object {
  "cells": Array [
    Array [
      Object {
        "cellId": 0,
        "letter": "S",
        "rowId": 0,
        "status": "ValidOutOfPosition",
      },
      Object {
        "cellId": 1,
        "letter": "W",
        "rowId": 0,
        "status": "Incorrect",
      },
      Object {
        "cellId": 2,
        "letter": "I",
        "rowId": 0,
        "status": "Incorrect",
      },
      Object {
        "cellId": 3,
        "letter": "N",
        "rowId": 0,
        "status": "Incorrect",
      },
      Object {
        "cellId": 4,
        "letter": "E",
        "rowId": 0,
        "status": "ValidOutOfPosition",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "letter": "S",
        "rowId": 1,
        "status": "ValidOutOfPosition",
      },
      Object {
        "cellId": 1,
        "letter": "T",
        "rowId": 1,
        "status": "ValidOutOfPosition",
      },
      Object {
        "cellId": 2,
        "letter": "R",
        "rowId": 1,
        "status": "ValidOutOfPosition",
      },
      Object {
        "cellId": 3,
        "letter": "I",
        "rowId": 1,
        "status": "Incorrect",
      },
      Object {
        "cellId": 4,
        "letter": "P",
        "rowId": 1,
        "status": "Incorrect",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "letter": "T",
        "rowId": 2,
        "status": "Correct",
      },
      Object {
        "cellId": 1,
        "letter": "R",
        "rowId": 2,
        "status": "Correct",
      },
      Object {
        "cellId": 2,
        "letter": "E",
        "rowId": 2,
        "status": "Correct",
      },
      Object {
        "cellId": 3,
        "letter": "E",
        "rowId": 2,
        "status": "Correct",
      },
      Object {
        "cellId": 4,
        "letter": "S",
        "rowId": 2,
        "status": "Correct",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 3,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 4,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 5,
        "status": "NotSet",
      },
    ],
  ],
  "focussed": Object {
    "cellId": -1,
    "rowId": -1,
    "status": "NotSet",
  },
  "gameComplete": true,
  "rowError": undefined,
  "success": true,
  "usedLetters": Object {
    "Correct": Set {},
    "Incorrect": Set {
      "W",
      "I",
      "N",
      "P",
    },
    "ValidOutOfPosition": Set {
      "S",
      "E",
      "T",
      "R",
    },
  },
}
`;

exports[`given game word with double letters given user adds valid word but not game word then hits enter should match snapshot 1`] = `
Object {
  "cells": Array [
    Array [
      Object {
        "cellId": 0,
        "letter": "T",
        "rowId": 0,
        "status": "Correct",
      },
      Object {
        "cellId": 1,
        "letter": "H",
        "rowId": 0,
        "status": "Incorrect",
      },
      Object {
        "cellId": 2,
        "letter": "E",
        "rowId": 0,
        "status": "Correct",
      },
      Object {
        "cellId": 3,
        "letter": "R",
        "rowId": 0,
        "status": "ValidOutOfPosition",
      },
      Object {
        "cellId": 4,
        "letter": "E",
        "rowId": 0,
        "status": "ValidOutOfPosition",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 1,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 2,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 3,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 4,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 5,
        "status": "NotSet",
      },
    ],
  ],
  "focussed": Object {
    "cellId": 0,
    "rowId": 1,
    "status": "NotSet",
  },
  "gameComplete": false,
  "rowError": undefined,
  "success": false,
  "usedLetters": Object {
    "Correct": Set {
      "T",
      "E",
    },
    "Incorrect": Set {
      "H",
    },
    "ValidOutOfPosition": Set {
      "R",
    },
  },
}
`;

exports[`updateGame tests given user adds invalid word then hits enter should match snapshot 1`] = `
Object {
  "cells": Array [
    Array [
      Object {
        "cellId": 0,
        "letter": "P",
        "rowId": 0,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "letter": "A",
        "rowId": 0,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "letter": "R",
        "rowId": 0,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "letter": "P",
        "rowId": 0,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "letter": "S",
        "rowId": 0,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 1,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 1,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 2,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 2,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 3,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 3,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 4,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 4,
        "status": "NotSet",
      },
    ],
    Array [
      Object {
        "cellId": 0,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 1,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 2,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 3,
        "rowId": 5,
        "status": "NotSet",
      },
      Object {
        "cellId": 4,
        "rowId": 5,
        "status": "NotSet",
      },
    ],
  ],
  "focussed": Object {
    "cellId": 4,
    "rowId": 0,
    "status": "NotSet",
  },
  "rowError": "Invalid word.",
  "usedLetters": Object {
    "Correct": Set {},
    "Incorrect": Set {},
    "ValidOutOfPosition": Set {},
  },
}
`;
